app/v1/asset/keeper.go

```diff
@@ -18,24 +21,200 @@ type Keeper struct {
+	if bz == nil {
+		return Gateway{}, ErrUnkwownGateway(k.codespace, fmt.Sprintf("Unknown gateway moniker:%s", moniker))
+	}
+
+	var gatewayID uint64
+	k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &gatewayID)
+
+	return k.GetGateway(ctx, gatewayID)
+}
+
+// HasGateway checks if the given gateway exists. Return true if exists, false otherwise
+func (k Keeper) HasGateway(ctx sdk.Context, moniker string) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Has(KeyMoniker(moniker))
+}
+
+// saveGateway saves the gateway, which behaves by the specified mode
+func (k Keeper) saveGateway(ctx sdk.Context, gateway Gateway, creation bool) {
+	k.setGateway(ctx, gateway)
+
+	if creation {
+		k.setMoniker(ctx, gateway.Moniker, gateway.ID)
+		k.setOwnerGatewayID(ctx, gateway.Owner, gateway.ID)
+		k.increaseOwnerGatewayCount(ctx, gateway.Owner)
+		k.increaseGatewayCount(ctx)
+	}
+}
+
+// setGateway stores the given gateway into underlying storage
+func (k Keeper) setGateway(ctx sdk.Context, gateway Gateway) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(gateway)
+
+	// set KeyGateway
+	store.Set(KeyGateway(gateway.ID), bz)
+}
+
+// setMoniker stores the gateway ID into storage by the key KeyMoniker
+func (k Keeper) setMoniker(ctx sdk.Context, moniker string, gatewayID uint64) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(gatewayID)
+
+	// set KeyMoniker
+	store.Set(KeyMoniker(moniker), bz)
+}
+
+// setOwnerGatewayID stores the gateway ID into storage by the key KeyOwnerGateway. Intended for iteration on ids of an owner
+func (k Keeper) setOwnerGatewayID(ctx sdk.Context, owner sdk.AccAddress, gatewayID uint64) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshalBinaryLengthPrefixed(gatewayID)
+
+	// set KeyOwnerGatewayID
+	store.Set(KeyOwnerGatewayID(owner, gatewayID), bz)
+}
+
+// increaseOwnerGatewayCount increases the count of the gateways of the given owner
+func (k Keeper) increaseOwnerGatewayCount(ctx sdk.Context, owner sdk.AccAddress) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(KeyOwnerGatewayCount(owner))
+
+	var count uint64
+	if bz == nil {
+		count = 0
+	} else {
+		k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &count)
+	}
+
+	bz = k.cdc.MustMarshalBinaryLengthPrefixed(count + 1)
+
+	// set KeyOwnerGatewayCount
+	store.Set(KeyOwnerGatewayCount(owner), bz)
+}
+
+// increaseGatewayCount increases the count of all the gateways
+func (k Keeper) increaseGatewayCount(ctx sdk.Context) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(KeyGatewayCount)
+
+	var count uint64
+	if bz == nil {
+		count = 0
+	} else {
+		k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &count)
+	}
+
+	bz = k.cdc.MustMarshalBinaryLengthPrefixed(count + 1)
+
+	// set KeyGatewayCount
+	store.Set(KeyGatewayCount, bz)
+}
+
+// getNewGatewayID gets the next available gateway ID and increments it
+func (k Keeper) getNewGatewayID(ctx sdk.Context) (gatewayID uint64, err sdk.Error) {
```

Comment:
`get` prefix of a method name is used for readonly methods, shall we use `generate` instead ?